### ISSUE TITLE
Feature: Load MegaLoc from bundled checkpoint

### ISSUE THREAD

**Issue by markvilar** [markvilar/viper#10]
## Problem and motivation

The current MegaLoc factory (`src/viper/models/megaloc.py`) constructs the model via `torch.hub.load("gmberton/MegaLoc", "get_trained_model")`, which pulls **both the model code and the trained weights from the external `gmberton/MegaLoc` repository at runtime**. This ties a core model to a third-party repository we don't control and introduces a hidden dependency that is only resolved at load time — the implementation lives outside this codebase, requires network access to fetch, and can change out from under us without any change on our side.

We want a **standalone MegaLoc implementation with model checkpoints owned by this repository**, independent of third-party repos and free of hidden runtime dependencies. Beyond the obvious reproducibility, offline-loadability, and provenance benefits, owning the implementation is a prerequisite for **decomposing and modifying the model**:

- To manipulate the model's internals we need the module classes to be first-class, readable, and importable within `viper` — not opaque code vendored in at runtime by `torch.hub`.
- In particular, the planned model-decomposition work (#6) performs principled truncation/decomposition of specific layers — e.g. SVD-based truncation of MegaLoc's `aggregator.linear` projection to reduce descriptor dimensionality. This requires stable, direct access to the model's layer structure and the ability to reconstitute it in new configurations, which a self-contained implementation makes tractable.

Bundling a fixed MegaLoc checkpoint and vendoring the module classes into a dedicated `megaloc/` package makes the model self-contained and fully under our control, and unlocks the downstream decomposition and modification work we intend to build on top of it.

## Proposed solution

Replace the `torch.hub`-based MegaLoc factory with a self-contained implementation vendored into `viper`, backed by a checkpoint hosted on this repository. Three parts: shared modules, the MegaLoc package, and checkpoint hosting + factory.

### 1. Shared modules — `viper.models.shared`

DINOv2 and SALAD are used by MegaLoc and other DINOv2/SALAD-based VPR models, so they are vendored into a new shared package rather than into the MegaLoc package. Copied from the reference implementation (`markvilar/benthloc-experiments`, `src/experiments/kappaplace/models/{dinov2,salad}.py`) and adapted to `viper` conventions, dropping KappaPlace-specific coupling.

- `viper/models/shared/dinov2.py` — `DINOv2` + sub-modules (`PatchEmbedding`, `Attention`, `LayerScale`, `Mlp`, `TransformerBlock`) and `build_dinov2_vitb14()`.
- `viper/models/shared/salad.py` — `SALAD`, `L2Norm`, the Sinkhorn/OTP helpers (`log_otp_solver`, `get_matching_probs`), and `build_salad(dropout=0.0)`.

### 2. MegaLoc components — `viper.models.megaloc`

`MegaLocModel` is composed of a backbone module and an aggregation module, each holding a shared module. Skeletons below convey the design; forward bodies elided.

**Backbone module** — thin adapter over `DINOv2`; resizes inputs to a patch-size multiple (DINOv2 tokenises on a fixed patch grid) and names the trunk output.

```python
class DinoFeatures(NamedTuple):
    patches: Tensor      # (B, C, H/14, W/14)
    cls_token: Tensor    # (B, C)

class MegaLocBackboneModule(nn.Module):
    """DINOv2 ViT-B/14 trunk -> patch grid + CLS token."""
    def __init__(self, dinov2: DINOv2) -> None:
        super().__init__()
        self.dinov2 = dinov2

    @property
    def num_channels(self) -> int:
        return self.dinov2.num_channels

    def forward(self, images: Tensor) -> DinoFeatures:
        images = resize_to_patch_multiple(images, self.dinov2.patch_size)
        patches, cls_token = self.dinov2(images)
        return DinoFeatures(patches, cls_token)
```

**Aggregation module** — holds the SALAD aggregator, the linear projection, and L2 normalisation. `aggregator.linear` (16640 -> 8448) is the exact layer issue #6 truncates.

```python
class MegaLocAggregationModule(nn.Module):
    """SALAD -> Linear -> L2Norm -> 8448-D descriptor."""
    def __init__(self, salad: SALAD, linear: nn.Linear) -> None:
        super().__init__()
        self.salad = salad
        self.linear = linear
        self.norm = L2Norm()

    @property
    def descriptor_dim(self) -> int:
        return self.linear.out_features      # 8448

    def forward(self, features: DinoFeatures) -> Tensor:
        return self.norm(self.linear(self.salad(features)))
```

**Model composite** — runs `backbone -> aggregator` and exposes the `ImageEmbedder` protocol surface. Grayscale->RGB handling is preserved from the current wrapper; `nn.Module.__call__` dispatches to `forward`.

```python
class MegaLocModel(nn.Module):
    """Composite MegaLoc embedder. Satisfies the ImageEmbedder protocol."""
    def __init__(self, backbone: MegaLocBackboneModule,
                 aggregator: MegaLocAggregationModule) -> None:
        super().__init__()
        self.backbone = backbone
        self.aggregator = aggregator

    @property
    def name(self) -> str:
        return "megaloc"

    @property
    def vector_size(self) -> int:
        return self.aggregator.descriptor_dim

    @property
    def embedder_parameters(self) -> dict[str, Any]:
        return {"backbone": "dinov2",
                "backbone_channels": self.backbone.num_channels}

    @property
    def device(self) -> str:
        return next(self.parameters()).device

    def forward(self, images: Tensor) -> Tensor:
        if images.shape[1] == 1:
            images = convert_grayscale_batch_to_rgb(images)
        assert images.dim() == 4, f"expected (B, C, H, W), got {images.shape}"
        assert images.shape[1] == 3, f"expected 3 channels, got {images.shape[1]}"
        return self.aggregator(self.backbone(images))
```

### 3. Checkpoint hosting, re-keying, and factory

Download the upstream raw checkpoint (`megaloc_raw.pth`, ~915 MB) from `markvilar/benthloc-experiments` release [`megaloc-raw-v1.0`](https://github.com/markvilar/benthloc-experiments/releases/tag/megaloc-raw-v1.0), re-key it (below), and upload as **`megaloc-8448d-v1.pth`** to a `viper` release (`8448d` = descriptor dimensionality, `v1` = model version tag).

The raw checkpoint's top-level layout already matches the composition (`backbone.*` / `aggregator.*`); only two inner names differ. We do a **one-time flat prefix rewrite** so the hosted checkpoint carries self-documenting names and loads with `strict=True` and no load-time remap — the names then serve as the interface for the #6 decomposition work.

| Original prefix | Re-keyed prefix | Why |
|---|---|---|
| `backbone.model.` | `backbone.dinov2.` | names the shared module the backbone holds |
| `aggregator.agg.` | `aggregator.salad.` | names the shared module the aggregator holds |
| `aggregator.linear.` | `aggregator.linear.` | already clear — unchanged |

```python
# one-time, run before upload; record source sha256 + this map in release notes
REMAP = {"backbone.model.": "backbone.dinov2.",
         "aggregator.agg.":  "aggregator.salad."}

def rekey(key: str) -> str:
    for old, new in REMAP.items():
        if key.startswith(old):
            return new + key[len(old):]
    return key
```

The factory then replaces `torch.hub.load(...)` with a checkpoint load, keeping the CUDA assertion (the DINOv2 backbone requires CUDA):

```python
_CHECKPOINT_URL = ".../releases/download/megaloc-8448d-v1/megaloc-8448d-v1.pth"

@register_embedder_factory(key="megaloc")
def load_megaloc() -> ImageEmbedder:
    if not torch.cuda.is_available():
        raise RuntimeError("CUDA is required for MegaLoc but is not available.")
    state = torch.hub.load_state_dict_from_url(
        _CHECKPOINT_URL, map_location="cpu", weights_only=True)
    model = MegaLocModel(
        backbone=MegaLocBackboneModule(build_dinov2_vitb14()),
        aggregator=MegaLocAggregationModule(
            salad=build_salad(dropout=0.0),
            linear=nn.Linear(16640, 8448)),
    )
    model.load_state_dict(state, strict=True)   # keys pre-aligned, no remap
    return model.eval().cuda()
```

## Open questions

None at this time.

## Notes

- **Checkpoint provenance.** `megaloc-8448d-v1.pth` is derived from the upstream raw checkpoint `megaloc_raw.pth` (release [`megaloc-raw-v1.0`](https://github.com/markvilar/benthloc-experiments/releases/tag/megaloc-raw-v1.0), `sha256:6279b37eb4bbcc894cf895d0e0bf858d5a7401506a793c2a0017ffc99086bbe5`) by the flat prefix re-key in §3 — weights are otherwise byte-for-byte unchanged. Record both hashes in the `viper` release notes.
- Module classes are copied from `markvilar/benthloc-experiments` (`src/experiments/kappaplace/models/`) as the starting point.
- Related: #6 (model decomposition) truncates `aggregator.linear` and depends on this standalone implementation.

*TBD: Populate section...*
